Sets : Union Supports Python set

souravgl0 opened the pull request:
Union Calls _sympify when an argument is a python set, except when
any element of that set is a sympy Set. This is because _sympify gives
unintended result when an element of set is a Set.
@asmeurer Does this solve the issue?
Example

    >>> Union({1},{2},{3})
    {1,2,3}
    >>> type(Union({1},{2},{3}))
    <class 'sympy.sets.sets.FiniteSet'>

However, This does not work.
        
    >>> Union({1,Interval(0,1)})
        TypeError: Input must be Sets or iterables of Sets

Adds Test in test_union() for checking this functionality.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #15448 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
    *  `Union` now supports Python set `Union({1,2},{3,4})={1,2,3,4}`

<!-- END RELEASE NOTES -->

souravgl0 committed:
Fixes #15548: Union Supports Python set

Union Calls _sympify when an argument is a python set, except when
any element of that set is a sympy Set. This is because _sympify gives
untinded result when an element of set is a Set.
Example
>>> Union({1},{2},{3})
{1,2,3}
>>> type(Union({1},{2},{3}))
<class 'sympy.sets.sets.FiniteSet'>

However, This does not work.
>>> Union({1,Interval(0,1)})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/sets/sets.py", line 1036, in __new__
    return S.EmptySet
  File "sympy/sets/sets.py", line 1034, in flatten
    # Union of no sets is EmptySet
  File "sympy/sets/sets.py", line 1034, in flatten
    # Union of no sets is EmptySet
  File "sympy/sets/sets.py", line 1035, in flatten
    if len(args) == 0:
TypeError: Input must be Sets or iterables of Sets

Adds Test in test_union() for checking this functionality.

asmeurer commented:
Actually it already does the right thing if the single argument is itself a set:

```py
>>> Union(FiniteSet(FiniteSet(1, 2), FiniteSet(2, 4)))
{{1, 2}, {2, 4}}
```

So I think `Union({FiniteSet(1, 2), FiniteSet(2, 4)})` should do the same thing. 

The behavior of Union on a `list` or `tuple` I guess can stay as is, since those aren't actually set objects (though IMHO it shouldn't be supported).

souravgl0 committed:
Union Works for any python set

Union calls _sympify if any arg is a python set.
symplify_union([Interval(0,1),Interval(2,2)]) was returning a python set
{Interval(0,1),2} which was being used to create a Union Object. This caused
Union() to create a FiniteSet which changes the whole meaning of input.
simplify_union now converts set to a list before calling Union, which Union supports.

souravgl0 commented:
This works now. Any python set can be used with Union().
The reason I excluded python sets containing Set was because those were leading to a test fail for the case Union(Interval(0,1),Interval(2,2)). That was happening because simplify_union called Union with a set which changes the meaning of argument to FiniteSet when it should have used a list (or any other container).

souravgl0 commented:
There might be other places where Union() is called with a set, assuming that it interprets the set as just a container like list. Now as we are changing set to mean FiniteSet, there might be other issues.

asmeurer commented:
Yes, it would be a backwards incompatible change. Which means we should discuss if it is worth it. I think it could be because it would make things mathematically consistent. The definition of a Union of a single set is that single set. Automatic denesting is inconsistent and makes it difficult to deal with sets of sets.

Actually Union goes even further than that and completely flattens its arguments:

```py
>>> Union([[FiniteSet(1, 2)], FiniteSet(3)])
{1, 2, 3}
```

I don't see any reason why it should do that. 

I would like to hear other people's thoughts on this.

souravgl0 commented:
At current point `{}` or `[]` are taken as just containers and are being flattened out when Union is called. We are already supporting nested sets. Just not with python sets or lists.

    >>> Union(FiniteSet(FiniteSet(FiniteSet(1, 2)), FiniteSet(3)))
    {{3},{{1,2}}

Only change this pull request brings is that anythin inside `{}` will mean a FiniteSet(). Python doesn't support sets of sets. So we can't nest with `{}`. If you think we should support nesting with `[]`, IMHO that should be a seperate issue. But I think that might be a useful feature. Although, I would have preferred nesting with `{}`

@asmeurer Why are you talking about Union of a single set, here? That is already supported without this pull request. And as far as I am able to understand, that's not the issue being resolved here. I'd be interested to know why you are talking about it.

I tried to see why the tests are failing and it has nothing to do with this commit. If we merge with the current master and run tests again, they are not failing.

I found something odd that I'm not able to wrap my head around. If we go checkout commit [5b55bdd](https://github.com/sympy/sympy/commit/5b55bddb1b9add4d35fcc4cea7098e07ff203826) when the pull request #154446 was merged and run tests, they are failing. How is that possible with Travis CI in place? 

Is there a feature to run tests with the code that comes after merging with master?

asmeurer commented:
`Union(single_set)` should just give `single_set`. That is consistent with the definition of a union. Implicit flattening is bad because it makes it hard or impossible to work with sets of sets, and in this case, it's completely unnecessary because one can always take a Union of an iterable using unpacking `Union(*iterable)`. 

Regarding the tests, there are some test failures in Azure on master. These will need to be resolved before we can merge this. I think @Upabjojr is working on it. Both Travis and Azure merge the latest master before running the tests.  I don't know how the test failures actually slipped into master, or why they only fail on Azure and not Travis. That is something worth investigating.

asmeurer commented:
I think the test failures in master have been fixed by https://github.com/sympy/sympy/pull/15489. They should work again if you push a new commit. Or someone can restart the build. I think sometimes Travis doesn't always merge the latest master for some reason, so that might be why they didn't fail here.

asmeurer commented:
And I should mention that I see no value in the "denesting" feature of Union. Is it used anywhere in the code (do tests fail if it is removed)?

Upabjojr commented:
The tests should work now if you restart them.

asmeurer commented:
@souravgl0 one of us has to restart the tests. But if you are going to push more changes anyway, it isn't necessary, because the tests will run again anyway when you push your commit.

asmeurer commented:
I support the change. Perhaps you should write to the mailing list to see if anyone else objects to it, since it is technically a backwards incompatible change.

souravgl0 committed:
Union does not flatten. Nested lists are converted to sets inside set

While using the Union constructor, the sets can be inside a list, an iterable or unpacked.
However, list inside the outermost list or iterable are taken to mean as another set.

>>> Union([{1},{2}])
{1, 2}
>>> Union([{1},[{2}]])
{1, {2}}

Also, if more than one arguments are passed unpacked, those containing a list are taken
to be a set.

>>> Union({1},[{2}])
{1, {2}}

>>> Union(FiniteSet(1),[FiniteSet(2),[FiniteSet(3)]])
{1,{2},{{3}}}

Changes a test that used old convention.

souravgl0 commented:
Added changes to stop denesting. The new behavior is as follows:

While using the Union constructor, the sets can be inside a list, an iterable or unpacked.
However, list inside the outermost list or iterable are taken to mean as another set.

```python
>>> Union([{1},{2}])
{1, 2}
>>> Union([{1},[{2}]])
{1, {2}}
```
Also, if more than one arguments are passed unpacked, those containing a list are taken
to be a set.
```python
>>> Union({1},[{2}])
{1, {2}}
```
```python
>>> Union(FiniteSet(1),[FiniteSet(2),[FiniteSet(3)]])
{1,{2},{{3}}}
```

souravgl0 committed:
Fixes solve_poly_inequality to be compatible with new convention of Union

Fixes tests
Solve_poly_inequality called Union constructor using the old convention.

souravgl0 commented:
@asmeurer I suppose you are waiting for people to post any possible concerns before merging this. Is a week's time enough or is the norm longer in such cases?

souravgl0 commented:
@asmeurer No one posted any concerns. Shall we merge this or wait for some more time?

asmeurer commented on sympy/solvers/inequalities.py:
```diff
@@ -126,7 +126,7 @@ def solve_poly_inequalities(polys):
     """
     from sympy import Union
-    return Union(*[solve_poly_inequality(*p) for p in polys])
+    return Union(*sum([solve_poly_inequality(*p) for p in polys],[]))
```
Using `sum` to combine lists is bad, because it has quadratic performance. You should be able to get the same thing with `Union(*[i for p in polys for i in solve_poly_inequality(*p)])`.

souravgl0 replied:
I didn't know that. I'll change it to your suggestion. But I picked up the syntax only from `Union()` constructor in `sets.py`. There `sum` is used along with `map`

souravgl0 commented on sympy/sets/sets.py:
```diff
@@ -1014,18 +1029,33 @@ def __new__(cls, *args, **kwargs):
                     return [arg]
             if iterable(arg):  # and not isinstance(arg, Set) (implicit)
-                return sum(map(flatten, arg), [])
+                return [FiniteSet(*sum(map(flatten, arg), []))]
```
Here, `sum` is used to merge lists. @asmeurer Is this situation different from the one where you said using `sum` gives quadratic performance, probable because of `map`?

asmeurer replied:
Using `sum` to merge lists always has quadratic performance, because it is essentially computing `(((list1 + list2) + list3) + list4) + ...)`. Each time a new list is created it has to allocate memory for the list and put the elements in it. On the other hand, using something like `[j for i in a for j in i]` only creates one list (Python uses doubling for allocation, so this is not an issue).

You can easily test this yourself if you do something like `a = [[1, 2] for i in range(10000)]` and compare the runtimes of `sum(a, [])` vs. `[j for i in a for j in i]`. For me it's 248 ms vs. 1.57 ms. And it becomes more drastic if you increase the number of lists. For 100000 lists, it's 27.6 s vs. 12.4 ms (a 2000x speed difference). 

Actually, this same sort of issue can also occur with SymPy expressions, which is why you should use `Add(*args)` instead of `sum(args)`. `sum` should only be used for numbers (integers and floats). The reason this issue doesn't happen with integers is because integers do not grow linearly in size with summation (adding `n` integers does not create an integer that requires `n` times as much memory to represent). 

Anyway, I thought we were removing flattening, so I'm unclear why this is still here. Is there still logic in Union that fully flattens its arguments?

souravgl0 replied:
No, There is no flattening. The name of the function was flatten originally and I didn't change it. This is required logic to convert arguments such as a Union, python Set, into a list. Basically, processing arguments and converting in a standard format.

Would changing the name of function to something else be better?

souravgl0 replied:
I understood the complexity issues with using `sum`. But, I was wondering why was it used in the first place if it is not optimal, here?

souravgl0 committed:
Improve time complexity for merge lists in test_enqualities.py

Used sum() to merge lists which runs in quadratic time
Changed to nested list comprehension

souravgl0 commented:
@asmeurer Please review.

asmeurer commented:
Sorry for not getting back on this. We should try to get it finished up and merged before the next release, since it has a backwards incompatible change.

jksuom commented:
I find it hard to perceive what exactly `flatten` is doing. I would expect that `args` is a tuple whose items describe sets. Each argument could be an instance of `Set`. Then each Python `set` can converted to a `FiniteSet`. But what to do with other Python objects? For iterable objects like lists, I think there are two possibilities. They can either be rejected or transformed to a `FiniteSet`. I don't think that they should be opened and their items treated as new arguments. The only case where denesting should be necessary is with Union instances. Their arguments should be appended to the list of original arguments. Then there should be no need for iterated denesting.

asmeurer commented:
I am working on a better fix for this.

asmeurer commented:
Here is my PR https://github.com/sympy/sympy/pull/16344. I have cleaned up both Union and Intersection, allowing Python sets as input and removing the automatic flattening. There are also several other fixes that were required along the way.
